### PR TITLE
feat: Add LazyLinear layer

### LINKED ISSUE #437: [Feature] Implement LazyLinear / Dense layer that can inference input dim of a linear layer
Both PyTorch and JAX supports this lazy linear layer that can automatically infer what will be the input_dim to initialize using the input tensor's last dimension, and user only needs to specify the output_dim (features). However, in mlx it looks like we currently do not have this feature, and we still need to explicitly specify the input_dim in mlx.nn.Linear

Related APIs
https://flax.readthedocs.io/en/latest/api_reference/flax.linen/_autosummary/flax.linen.Dense.html
https://pytorch.org/docs/stable/generated/torch.nn.LazyLinear.html#torch.nn.LazyLinear

### PR DESCRIPTION
**Pull request by NripeshN**
## Proposed changes

Added lazy linear
Closes #437

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by NripeshN** [comment]
@awni review me please😅

**Comment by awni** [comment]
Sorry for the delay, will try and get to this tomorrow 🤞

**Comment by awni** [comment]
I have some doubts about how this will work. For example what if I get `module.parameters` before the network forward has been called? If I want do e.g. a gradient transform then the weight will be missing from the params right?

```python
def loss(params, input, target);
   model.update(params)
   y = model(input)
   return nn.losses.mse(y, target)

grads = mx.grad(loss)(module.parameters(), input, target)
```

I don't think `grads` would have the weight on the first iteration right? I find that a bit strange...but I'm not sure if there is a reasonable workaround

**Comment by awni** [comment]
@NripeshN what do you think about my comment? If we don't have a clean solution for it I'm inclined not to add this module as the usage can lead to unexpected / surprising results.

**Comment by NripeshN** [comment]
> I have some doubts about how this will work. For example what if I get `module.parameters` before the network forward has been called? If I want do e.g. a gradient transform then the weight will be missing from the params right?
> 
> ```python
> def loss(params, input, target);
>    model.update(params)
>    y = model(input)
>    return nn.losses.mse(y, target)
> 
> grads = mx.grad(loss)(module.parameters(), input, target)
> ```
> 
> I don't think `grads` would have the weight on the first iteration right? I find that a bit strange...but I'm not sure if there is a reasonable workaround

You've raised a valid point regarding the behavior of the `LazyLinear` module, particularly concerning the scenario where `module.parameters()` is called before the first forward pass. Indeed, in its current form, the weights and biases of the `LazyLinear` module are not initialized until the first forward pass. As a result, if you try to access `module.parameters()` or compute gradients before this happens, the weights and biases won't be present, leading to the issue you've mentioned.

This behavior is a fundamental aspect of lazy initialization: parameters are not created until they're needed, which is both a feature and a limitation depending on the use case. In the standard eager initialization approach, all parameters are initialized when the module is created, making them immediately available for operations like gradient computation.

However, there are a few ways to address this in the context of lazy initialization:

1. **Explicit Initialization Call**: We could introduce an explicit method to initialize weights and biases. This way, you could manually initialize them before performing operations like gradient computation. This does slightly deviate from the "lazy" aspect, as it requires manual intervention.

2. **Parameter Placeholder**: We could implement a placeholder mechanism for parameters. When the module is instantiated, placeholders for weights and biases could be created and included in `module.parameters()`. These would be replaced with actual parameters upon the first forward pass. This approach ensures that `module.parameters()` always returns something, but it adds complexity and might lead to confusion if the placeholders are mistakenly used in computations.

For your specific use case, it's important to ensure that the first forward pass occurs before attempting operations that rely on the existence of these parameters. This is a peculiarity of using lazy initialization, and while it offers flexibility in some scenarios (like dynamic input sizes), it does require careful handling in others.

**Comment by awni** [comment]
Yea I don't love either of those options. The first one is not that bad (e.g. forcing an explicit initialization call). The problem is we can't know the sizes until an actually forward pass has been done. So the initialization call would require inputs.

In fact just doing `model(inputs)` and ignoring the results I think is sufficient for an initialization.

But at the moment, I think the extra complexity of using this somewhat defeats the purpose... and specifying shapes in regular Linear layers is not such a hardship. Maybe others feel differently, but I'm leaning towards not including this until we see a strong need for it.

**Comment by awni** [comment]
@NripeshN any further comments on this? Should we close for now and revisit if the need arises?

**Comment by NripeshN** [comment]
> @NripeshN any further comments on this? Should we close for now and revisit if the need arises?

Yup we can close this PR for now. We could reopen it when it is necessary.
Thanks!
